emukit/bayesian_optimization/acquisitions/expected_improvement.py

```diff
@@ -88,3 +90,338 @@ def get_standard_normal_pdf_cdf(x: np.array, mean: np.array, standard_deviation:
+                    B[l, j] = 1.0 / eps * (f1 - f)
+                grad = grad + B
+            else:
+                B1, B2, B3 = np.zeros((q, d)), np.zeros((q, d)), np.zeros((q, d))
+                for i in range(q):
+                    ineq = [n for n in range(q) if n is not i]
+                    Sigk_ik = Sigk[i, k]
+                    Sigk_ii = Sigk[i, i]
+                    mk_i = mk[i]
+                    mk_dx_i = mk_dx[:, :, i]
+                    bk_i = bk[i, 0]
+                    ck_pi = (bk[ineq, 0] - mk[ineq, 0] - (bk[i, 0] - mk[i, 0]) / Sigk_ii * Sigk[ineq, :][:, i])[:, None]
+                    Sigk_pi = 0.5 * (Sigk[ineq, :][:, ineq]
+                                     - 1.0 / Sigk_ii * Sigk[ineq, :][:, i, None] @ Sigk[ineq, :][:, i, None].T
+                                     + (Sigk[ineq, :][:, ineq]
+                                     - 1.0 / Sigk_ii * Sigk[ineq, :][:, i, None] @ Sigk[ineq, :][:, i, None].T).T)
+                    Sigk_dx_ii = Sigk_dx[:, :, i, i]
+                    Sigk_dx_ik = Sigk_dx[:, :, i, k]
+                    phi_ik = np.max([scipy.stats.multivariate_normal.pdf(bk[i, 0], mk_i, Sigk_ii), 1e-11])
+                    dphi_ik_dSig = ((bk_i - mk_i)**2 / (2.0 * Sigk_ii**2) - 0.5 / Sigk_ii) * phi_ik
+                    dphi_ik_dm = (bk_i - mk_i) / Sigk_ii * phi_ik
+                    Phi_ik = symmetric_term[k, i] / phi_ik
+                    GPhi_ik = GPhi(ck_pi, np.zeros((q - 1, 1)), Sigk_pi)
+                    HPhi_ik = HPhi(ck_pi, np.zeros((q - 1, 1)), Sigk_pi, GPhi_ik)
+                    Sigk_mi = Sigk[ineq, i, None]
+                    for l, j in itertools.product(range(q), range(d)):
+                        # B1
+                        B1[l, j] = B1[l, j] + Sigk_dx_ik[l, j] * phi_ik * Phi_ik
+                        # B2
+                        B2[l, j] = B2[l, j] + Sigk_ik * (mk_dx_i[l, j] * dphi_ik_dm
+                                                         + dphi_ik_dSig * Sigk_dx_ii[l, j]) * Phi_ik
+                        # B3
+                        dck_pi = (-mk_dx[l, j, ineq] + (mk_dx_i[l, j] * Sigk_ii
+                                                        + (bk_i - mk_i) * Sigk_dx_ii[l, j])
+                                  / (Sigk_ii**2) * Sigk_mi[:, 0]
+                                  - (bk[i, 0] - mk_i) / Sigk_ii * Sigk_dx[l, j, ineq, i])
+                        SigtCross = Sigk_dx[l, j, ineq, i, None] @ Sigk_mi.T
+                        dSigk_pi = (Sigk_dx[l, j, :, :][:, ineq][ineq, :]
+                                    + Sigk_dx_ii[l, j] / (Sigk_ii**2) * (Sigk_mi @ Sigk_mi.T)
+                                    - (1.0 / Sigk_ii) * (SigtCross + SigtCross.T))
+                        B3[l, j] = (B3[l, j] + Sigk_ik * phi_ik * ((GPhi_ik.T @ dck_pi)
+                                    + 0.5 * np.sum(HPhi_ik * dSigk_pi)))
+                grad = grad + B1 + B2 + B3
+        return grad
+
+
+def covZk(Sigma: np.ndarray, k: int) -> np.ndarray:
```

Comment:
Another point about these helper methods. Let's clearly indicate them as private with leading underscores - i don't think they can be reused for anything else

Reply:
These helper methods are not inside the class as they are not using any class attributes. In order to indicate that they are private (with underscore), I would have to move them inside the class. I kinda made them private by only importing the class in the __init__.py-file, so that these functions are not visible if not importing this file directly.

If I would move them inside the class, it would be natural to define them as @staticmethod but then I couldn't make them private (Or could I?). Is there an elegant way to define private static methods? To my naive understanding, the current way is the closest I can get to it.

I agree that no one probably needs these methods anywhere, but on the other hand they are generic, but rarely used tricks to condition a multivariate normal distribution.

What do you think?

Reply:
thanks for explanation. i think it's fine as it is now, no need for more changes!